client/src/nonblocking/blockhash_query.rs

```diff
@@ -0,0 +1,433 @@
+use {
+    crate::nonblocking::{nonce_utils, rpc_client::RpcClient},
+    clap::ArgMatches,
+    solana_clap_utils::{
+        input_parsers::{pubkey_of, value_of},
+        nonce::*,
+        offline::*,
+    },
+    solana_sdk::{commitment_config::CommitmentConfig, hash::Hash, pubkey::Pubkey},
+};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Source {
+    Cluster,
+    NonceAccount(Pubkey),
+}
+
+impl Source {
+    pub async fn get_blockhash(
+        &self,
+        rpc_client: &RpcClient,
+        commitment: CommitmentConfig,
+    ) -> Result<Hash, Box<dyn std::error::Error>> {
+        match self {
+            Self::Cluster => {
+                let (blockhash, _) = rpc_client
+                    .get_latest_blockhash_with_commitment(commitment)
+                    .await?;
+                Ok(blockhash)
+            }
+            Self::NonceAccount(ref pubkey) => {
+                #[allow(clippy::redundant_closure)]
+                let data = nonce_utils::get_account_with_commitment(rpc_client, pubkey, commitment)
+                    .await
+                    .and_then(|ref a| nonce_utils::data_from_account(a))?;
+                Ok(data.blockhash())
+            }
+        }
+    }
+
+    pub async fn is_blockhash_valid(
+        &self,
+        rpc_client: &RpcClient,
+        blockhash: &Hash,
+        commitment: CommitmentConfig,
+    ) -> Result<bool, Box<dyn std::error::Error>> {
+        Ok(match self {
+            Self::Cluster => rpc_client.is_blockhash_valid(blockhash, commitment).await?,
```

Comment:
Now that I'm actually reading this, I'm remembering this discussion: https://github.com/solana-labs/solana/issues/23949#issuecomment-1086408022

Let's stick with this for now, but can you create an issue to clean this up and *not* use `rpc_client.is_blockhash_valid`?

Reply:
#27045